src/gazebo_lidar_plugin.cpp

```diff
@@ -108,6 +109,84 @@ void RayPlugin::Load(sensors::SensorPtr _parent, sdf::ElementPtr _sdf)
+                              [](const string& name)
+                              { return name.size() == 0; }), end(names_splitted));
+
+  std::string rootModelName = names_splitted.front(); // The first element is the name of the root model
+
+  // Get the pointer to the root model
+#if GAZEBO_MAJOR_VERSION >= 9
+  physics::ModelPtr rootModel = this->world->ModelByName(rootModelName);
+#else
+  physics::ModelPtr rootModel = this->world->GetModel(rootModelName);
+#endif
+
+  // Get the `base_link` rotation WRT world
+  physics::LinkPtr baseLink = nullptr;
+  std::vector<physics::LinkPtr> linkList = rootModel->GetLinks(); // Get list of all links in the root model
+  for(auto link : linkList) {
+    std::string linkName = link->GetName();
+    if(linkName.find("::base_link") != std::string::npos) {
+      baseLink = rootModel->GetLink(linkName); // Get the pointer to the `base_link`
+      break;
+    }
+  }
+  if (!baseLink)
+    gzthrow("RayPlugin requires the `base_link` element to be defined");
+
+  // This is the rotation of the 'base_link` WRT world
+#if GAZEBO_MAJOR_VERSION >= 9
+  ignition::math::Quaterniond q_wb = baseLink->WorldPose().Rot();
+#else
+  ignition::math::Quaterniond q_wb = ignitionFromGazeboMath(baseLink->GetWorldPose()).Rot();
+#endif
+
+  // Get the parent sensor link rotation WRT world
+  physics::LinkPtr parentSensorLink = nullptr;
+  std::string parentSensorModelName = names_splitted.rbegin()[1]; // the second to the last name is the model name
+  for(auto link : linkList) {
+    std::string linkName = link->GetName();
+    if(linkName.find(parentSensorModelName) != std::string::npos) {
+      parentSensorLink = rootModel->GetLink(linkName); // Get the pointer to the parent sensor link
+      break;
+    }
+  }
+  if (!parentSensorLink)
+    gzthrow("RayPlugin requires a `link` element for its parent sensor to be defined");
+
+  // This is the rotation of the parentSensorLink WRT world
+#if GAZEBO_MAJOR_VERSION >= 9
+  ignition::math::Quaterniond q_wl = parentSensorLink->WorldPose().Rot();
+#else
+  ignition::math::Quaterniond q_wl = ignitionFromGazeboMath(parentSensorLink->GetWorldPose()).Rot();
+#endif
+
+  // Calculate parent sensor rotation WRT `base_link`
+  ignition::math::Quaterniond q_bl = q_wb.Inverse() * q_wl; // This is the rotation of the parent sensor link WRT `base_link`
+
+  ignition::math::Quaterniond q_ls = parentSensor_->Pose().Rot(); // This is the rotation of the parent sensor WRT parent sensor link
+  ignition::math::Quaterniond q_bs = (q_bl * q_ls).Inverse(); // This is the rotation of the parent sensor WRT `base_link`
+
+  const ignition::math::Vector3d u_Xb = kForwardRotation; // This is unit vector of X-axis `base_link`
+  const ignition::math::Vector3d u_Xs = q_bs.RotateVectorReverse(u_Xb); // This is unit vector of X-axis sensor in `base_link` frame
+
+  // Current rotation types are described as https://github.com/PX4/Firmware/blob/master/msg/distance_sensor.msg
```

Comment:
Can you make this Mavlink compliant instead?

Reply:
Done if i got you right.

Reply:
@TSC21 good enough?

Reply:
@TSC21 Is it worth to proceed with the sonar plugin?